In [76]:
import pandas as pd
import sqlite3

In [77]:
master_df = pd.read_csv("../data/cleaned/olist_master_cleaned.csv")

master_df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,review_creation_date,review_answer_timestamp,delivery_days,delivery_delay,total_order_value,shipping_ratio,order_month,order_year,order_quarter,purchase_weekday
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
1,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
2,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
3,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,...,2018-08-08 00:00:00,2018-08-08 18:37:50,13.0,-6.0,141.46,0.191744,July,2018,Q3,Tuesday
4,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,2018-08-18 00:00:00,2018-08-22 19:07:58,9.0,-18.0,179.12,0.120200,August,2018,Q3,Wednesday


In [78]:
connection = sqlite3.connect("../data/retail_pricing.db")

print("Database created successfully!")

Database created successfully!


In [79]:
master_df.to_sql(
    name="olist_master",
    con=connection,
    if_exists="replace",
    index=False
)

print("Table 'olist_master' created successfully!")

Table 'olist_master' created successfully!


In [80]:
pd.read_sql(
    """
    SELECT name
    FROM sqlite_master
    WHERE type='table';
    """,
    connection
)

,name
0,olist_master


# SQL Analysis 1: Dataset Overview

## Business Question

How large is the sales dataset and what information does it contain?

---

## Business Importance

Before performing any business analysis, analysts first validate the dataset.

This helps answer questions such as:

- How many records exist?
- How many variables are available?
- Is the data ready for analysis?

Dataset validation is the first step of every real-world SQL workflow.

---

## SQL Queries

- SELECT
- COUNT()
- LIMIT

In [81]:
pd.read_sql(
    """
    SELECT *
    FROM olist_master
    LIMIT 5;
    """,
    connection
)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,review_creation_date,review_answer_timestamp,delivery_days,delivery_delay,total_order_value,shipping_ratio,order_month,order_year,order_quarter,purchase_weekday
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
1,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
2,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
3,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,...,2018-08-08 00:00:00,2018-08-08 18:37:50,13.0,-6.0,141.46,0.191744,July,2018,Q3,Tuesday
4,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,2018-08-18 00:00:00,2018-08-22 19:07:58,9.0,-18.0,179.12,0.120200,August,2018,Q3,Wednesday


In [82]:
pd.read_sql(
    """
    SELECT COUNT(*) AS total_rows
    FROM olist_master;
    """,
    connection
)

,total_rows
0,119143


## 📦 Business Insight

The SQL database has been successfully created and contains the complete retail sales dataset.

The dataset is now ready for business reporting, KPI calculations, and advanced SQL analytics.

# SQL Analysis 2: Business KPIs

## Business Question

What are the overall sales performance metrics of the marketplace?

---

## Business Importance

Business leaders rely on high-level KPIs to evaluate marketplace performance.

These metrics provide a quick snapshot of:

- Total Revenue
- Total Orders
- Average Order Value
- Customer Base

These KPIs are typically displayed on executive dashboards and are the foundation of business decision-making.

---

## SQL Queries

- COUNT()
- SUM()
- AVG()
- ROUND()
- DISTINCT

In [83]:
pd.read_sql(
    """
    SELECT
        ROUND(SUM(payment_value), 2) AS total_revenue
    FROM olist_master;
    """,
    connection
)

,total_revenue
0,20579664.01


In [84]:
pd.read_sql(
    """
    SELECT
        COUNT(DISTINCT order_id) AS total_orders
    FROM olist_master;
    """,
    connection
)

,total_orders
0,99441


In [85]:
pd.read_sql(
    """
    SELECT
        COUNT(DISTINCT customer_unique_id) AS total_customers
    FROM olist_master;
    """,
    connection
)

,total_customers
0,96096


In [86]:
pd.read_sql(
    """
    SELECT
        ROUND(AVG(payment_value), 2) AS average_order_value
    FROM olist_master;
    """,
    connection
)

,average_order_value
0,172.74


# SQL Analysis 3: Sales Performance by Order Status

## Business Question

How are orders distributed across different order statuses?

---

## Business Importance

Understanding order status helps businesses monitor operational efficiency and customer satisfaction.

Businesses can identify:

- Successfully delivered orders
- Cancelled orders
- Processing bottlenecks
- Customer fulfillment performance

Monitoring these metrics helps improve logistics, inventory management, and overall customer experience.

---

## SQL queries

- GROUP BY
- COUNT()
- ORDER BY

In [87]:
pd.read_sql(
    """
    SELECT
        order_status,
        COUNT(DISTINCT order_id) AS total_orders
    FROM olist_master
    GROUP BY order_status
    ORDER BY total_orders DESC;
    """,
    connection
)

,order_status,total_orders
0,delivered,96478
1,shipped,1107
2,canceled,625
3,unavailable,609
4,invoiced,314
5,processing,301
6,created,5
7,approved,2


In [88]:
pd.read_sql(
    """
    SELECT
        order_status,
        COUNT(DISTINCT order_id) AS total_orders,
        ROUND(
            COUNT(DISTINCT order_id) * 100.0 /
            (SELECT COUNT(DISTINCT order_id) FROM olist_master),
            2
        ) AS percentage
    FROM olist_master
    GROUP BY order_status
    ORDER BY total_orders DESC;
    """,
    connection
)

,order_status,total_orders,percentage
0,delivered,96478,97.02
1,shipped,1107,1.11
2,canceled,625,0.63
3,unavailable,609,0.61
4,invoiced,314,0.32
5,processing,301,0.30
6,created,5,0.01
7,approved,2,0.00


## 📦 Business Insight

The majority of marketplace orders are expected to be successfully delivered, indicating healthy fulfillment operations.

Monitoring smaller categories such as cancelled, unavailable, or invoiced orders can help identify operational inefficiencies and improve customer satisfaction.

Regularly tracking order status distribution enables businesses to reduce failed deliveries and optimize the order fulfillment process.

# SQL Analysis 4: Revenue by Product Category

## Business Question

Which product categories generate the highest revenue?

---

## Business Importance

Product category analysis helps businesses understand where revenue comes from.

These insights enable businesses to:

- Identify top-performing product categories.
- Allocate marketing budgets effectively.
- Optimize inventory planning.
- Improve pricing strategies.
- Focus on high-value product segments.

---

## SQL Queries

- GROUP BY
- SUM()
- ROUND()
- ORDER BY
- LIMIT

In [89]:
pd.read_sql("""
PRAGMA table_info(olist_master);
""", connection)

,cid,name,type,notnull,dflt_value,pk
0,0,order_id,TEXT,0,None,0
1,1,customer_id,TEXT,0,None,0
2,2,order_status,TEXT,0,None,0
3,3,order_purchase_timestamp,TEXT,0,None,0
4,4,order_approved_at,TEXT,0,None,0
5,5,order_delivered_carrier_date,TEXT,0,None,0
6,6,order_delivered_customer_date,TEXT,0,None,0
7,7,order_estimated_delivery_date,TEXT,0,None,0
8,8,customer_unique_id,TEXT,0,None,0
9,9,customer_zip_code_prefix,INTEGER,0,None,0


In [90]:
pd.read_sql("""
SELECT *
FROM olist_master
LIMIT 1;
""", connection).columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'customer_unique_id', 'customer_zip_code_prefix', 'customer_city',
       'customer_state', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value',
       'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'payment_sequential', 'payment_type', 'payment_installments',
       'payment_value', 'review_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'delivery_days', 'delivery_delay',
       'total_order_value', 'shipping_ratio', 'order_month', 'order_year',
       'order_quarter', 'purchase_weekda

In [91]:
pd.read_sql("""
SELECT
    product_category_name,
    ROUND(SUM(payment_value), 2) AS total_revenue
FROM olist_master
GROUP BY product_category_name
ORDER BY total_revenue DESC
LIMIT 10;
""", connection)

,product_category_name,total_revenue
0,cama_mesa_banho,1743998.80
1,beleza_saude,1662963.59
2,informatica_acessorios,1599481.06
3,moveis_decoracao,1443963.61
4,relogios_presentes,1430553.48
5,esporte_lazer,1400223.07
6,utilidades_domesticas,1097900.09
7,automotivo,855095.68
8,ferramentas_jardim,840721.59
9,cool_stuff,781933.97


In [92]:
pd.read_sql("""
SELECT
    customer_unique_id,
    ROUND(SUM(payment_value), 2) AS lifetime_value
FROM olist_master
GROUP BY customer_unique_id
ORDER BY lifetime_value DESC
LIMIT 10;
""", connection)

,customer_unique_id,lifetime_value
0,0a0a92112bd4c708ca5fde585afaa872,109312.64
1,698e1cf81d01a3d389d96145f7fa6df8,45256.00
2,c402f431464c72e27330a67f7b94d4fb,44048.00
3,4007669dec559734d6f53e029e360987,36489.24
4,ef8d54b3797ea4db1d63f0ced6a906e9,30186.00
5,763c8b1c9c68a0229c42c9fc6f662b93,29099.52
6,c8460e4251689ba205045f3ea17884a1,27935.46
7,eae0a83d752b1dd32697e0e7b4221656,25051.89
8,3d47f4368ccc8e1bb4c4a12dbda7111b,22346.60
9,adfa1cab2b2c8706db21bb13c0a1beb1,19457.04


In [93]:
pd.read_sql("""
SELECT
    strftime('%Y-%m', order_purchase_timestamp) AS purchase_month,
    ROUND(SUM(payment_value), 2) AS monthly_revenue
FROM olist_master
GROUP BY purchase_month
ORDER BY purchase_month;
""", connection)

,purchase_month,monthly_revenue
0,2016-09,388.47
1,2016-10,76559.05
2,2016-12,19.62
3,2017-01,190806.27
4,2017-02,351848.13
5,2017-03,547769.84
6,2017-04,512126.52
7,2017-05,737425.31
8,2017-06,613777.41
9,2017-07,749242.84


In [94]:
pd.read_sql("""
SELECT
    seller_id,
    ROUND(SUM(payment_value), 2) AS total_revenue
FROM olist_master
GROUP BY seller_id
ORDER BY total_revenue DESC
LIMIT 10;
""", connection)

,seller_id,total_revenue
0,7c67e1448b00f6e969d365cea6b010ab,512645.19
1,1025f0e2d44d7041d6cf58b6550e0bfa,312456.49
2,4a3ca9315b744ce9f8e9374361493884,306138.80
3,1f50f920176fa81dab994f9023523100,291918.98
4,53243585a1d6dc2643021fd1853d8905,284903.08
5,da8622b14eb17ae2831f4ac5b9dab84a,276578.63
6,4869f7a5dfa277a7dca6462dcf3b52b2,264166.12
7,955fee9216a65b617aa5c0531780ce60,236414.48
8,fa1c13f2614d7b5c4749cbc52fecda94,206513.23
9,7e93a43ef30c4f03f38b393420bc753a,185134.21


In [95]:
pd.read_sql("""
SELECT
    product_category_name,
    ROUND(AVG(review_score), 2) AS average_review_score
FROM olist_master
WHERE review_score IS NOT NULL
GROUP BY product_category_name
ORDER BY average_review_score DESC
LIMIT 10;
""", connection)

,product_category_name,average_review_score
0,cds_dvds_musicais,4.64
1,fashion_roupa_infanto_juvenil,4.50
2,livros_interesse_geral,4.44
3,livros_importados,4.42
4,flores,4.42
5,construcao_ferramentas_ferramentas,4.42
6,livros_tecnicos,4.38
7,portateis_casa_forno_e_cafe,4.32
8,alimentos_bebidas,4.32
9,malas_acessorios,4.30


In [96]:
pd.read_sql("""
SELECT
    payment_type,
    ROUND(SUM(payment_value), 2) AS total_revenue,
    COUNT(order_id) AS total_orders
FROM olist_master
GROUP BY payment_type
ORDER BY total_revenue DESC;
""", connection)

,payment_type,total_revenue,total_orders
0,credit_card,15775450.54,87776
1,boleto,4110920.74,23190
2,voucher,435917.84,6465
3,debit_card,257374.89,1706
4,not_defined,0.00,3
5,NaN,NaN,3


In [97]:
pd.read_sql("""
SELECT
    order_id,
    payment_type,
    payment_value
FROM olist_master
WHERE payment_type IS NULL
   OR payment_type = 'not_defined';
""", connection)

,order_id,payment_type,payment_value
0,00b1cb0320190ca0daa2c88b35206009,not_defined,0.0
1,bfbd0f9bdef84302105ad712db648a6c,NaN,NaN
2,bfbd0f9bdef84302105ad712db648a6c,NaN,NaN
3,bfbd0f9bdef84302105ad712db648a6c,NaN,NaN
4,4637ca194b6387e2d538dc89b124b0ee,not_defined,0.0
5,c8c528189310eaa44a745b8d9d26908b,not_defined,0.0


### 🔍 Observation: Payment Data Quality

During the payment method analysis, six records were identified with invalid or incomplete payment information.

- Three records had `payment_type = 'not_defined'` with a `payment_value` of `0.0`.
- Three records contained `NULL` values for both `payment_type` and `payment_value`.

These records account for less than **0.01%** of the dataset and are considered minor data quality issues. Due to their negligible impact on the overall analysis, they were excluded from business interpretation while being documented for transparency.

In [98]:
pd.read_sql("""
SELECT
    strftime('%Y-%m', order_purchase_timestamp) AS purchase_month,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master
GROUP BY purchase_month
ORDER BY purchase_month;
""", connection)

,purchase_month,total_orders
0,2016-09,4
1,2016-10,324
2,2016-12,1
3,2017-01,800
4,2017-02,1780
5,2017-03,2682
6,2017-04,2404
7,2017-05,3700
8,2017-06,3245
9,2017-07,4026


In [99]:
pd.read_sql("""
SELECT
    customer_unique_id,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master
GROUP BY customer_unique_id
ORDER BY total_orders DESC
LIMIT 10;
""", connection)

,customer_unique_id,total_orders
0,8d50f5eadf50201ccdcedfb9e2ac8455,17
1,3e43e6105506432c953e165fb2acf44c,9
2,ca77025e7201e3b30c44b472ff346268,7
3,6469f99c1f9dfae7733b25662e7f1782,7
4,1b6c7548a2a1f9037c1fd3ddfed95f33,7
5,f0e310a6839dce9de1638e0fe5ab282a,6
6,de34b16117594161a6a89c50b289d35a,6
7,dc813062e0fc23409cd255f7f53c7074,6
8,63cfc61cee11cbe306bff5857d00bfe4,6
9,47c1a3033b8b77b3ab6e109eb4d5fdf3,6


In [100]:
pd.read_sql("""
SELECT
    order_status,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master
GROUP BY order_status
ORDER BY total_orders DESC;
""", connection)

,order_status,total_orders
0,delivered,96478
1,shipped,1107
2,canceled,625
3,unavailable,609
4,invoiced,314
5,processing,301
6,created,5
7,approved,2


In [101]:
pd.read_sql("""
SELECT
    customer_state,
    ROUND(SUM(payment_value), 2) AS total_revenue,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master
GROUP BY customer_state
ORDER BY total_revenue DESC;
""", connection)

,customer_state,total_revenue,total_orders
0,SP,7726078.35,41746
1,RJ,2795615.67,12852
2,MG,2351221.09,11635
3,RS,1160175.66,5466
4,PR,1079795.49,5045
5,BA,805070.98,3380
6,SC,801276.45,3637
7,GO,520481.65,2020
8,DF,438095.32,2140
9,ES,408611.64,2033


In [102]:
pd.read_sql("""
SELECT
    customer_state,
    ROUND(AVG(delivery_days), 2) AS avg_delivery_days,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master
WHERE delivery_days IS NOT NULL
GROUP BY customer_state
ORDER BY avg_delivery_days DESC;
""", connection)

,customer_state,avg_delivery_days,total_orders
0,RR,27.83,41
1,AP,27.66,67
2,AM,26.06,145
3,AL,24.13,397
4,PA,23.27,946
5,MA,21.19,717
6,SE,20.89,335
7,CE,20.34,1279
8,AC,20.17,80
9,PB,20.11,517


In [103]:
pd.read_sql("""
SELECT
    CASE
        WHEN delivery_delay > 0 THEN 'Delayed'
        ELSE 'On Time'
    END AS delivery_status,
    COUNT(DISTINCT order_id) AS total_orders,
    ROUND(AVG(delivery_delay), 2) AS avg_delay_days
FROM olist_master
WHERE delivery_delay IS NOT NULL
GROUP BY delivery_status;
""", connection)

,delivery_status,total_orders,avg_delay_days
0,Delayed,6535,10.57
1,On Time,89941,-13.63


### Observation

Delivery performance is strong across the marketplace.

- 89,941 orders (~93%) were delivered on or before the estimated delivery date.
- 6,535 orders (~7%) were delivered late.
- Delayed orders arrived an average of **10.57 days** after the promised delivery date.
- Orders classified as "On Time" were delivered an average of **13.63 days before** the estimated delivery date, indicating that estimated delivery dates were generally conservative.

Overall, the logistics network appears reliable, with the majority of customers receiving their orders earlier than expected.

In [104]:
pd.read_sql("""
SELECT
    customer_unique_id,
    COUNT(DISTINCT order_id) AS total_orders,
    ROUND(SUM(payment_value), 2) AS lifetime_value,
    ROUND(AVG(payment_value), 2) AS avg_order_value
FROM olist_master
GROUP BY customer_unique_id
HAVING total_orders > 1
ORDER BY lifetime_value DESC
LIMIT 10;
""", connection)

,customer_unique_id,total_orders,lifetime_value,avg_order_value
0,c8460e4251689ba205045f3ea17884a1,4,27935.46,1163.98
1,eae0a83d752b1dd32697e0e7b4221656,2,25051.89,1670.13
2,0f5ac8d5c31de21d2f25e24be15bbffb,2,19342.26,1074.57
3,d97b3cfb22b0d6b25ac9ed4e9c2d481b,2,19167.26,798.64
4,906a8a4ec9f3d4c3e64fa6d1c4fe6009,2,11881.01,1697.29
5,5a494c648fde2d1ec4eb614274ea7159,2,9859.88,896.35
6,86df00dc5fd68f4dd5d5945ca19f3ed6,3,9601.92,800.16
7,46450c74a0d8c5ca9395da1daac6c120,3,9553.02,3184.34
8,efce1ab3e96ccab8b1b464326bd22417,2,7971.88,724.72
9,3ec83b54e0afb70def9480cd6b9097f7,2,7723.80,858.20


### Observation

The highest-value repeat customers generated substantial revenue despite making relatively few purchases.

- The top customer placed only 4 orders but generated a lifetime value of 27,935.46.
- Most customers in the top 10 made just 2 to 4 purchases, indicating that high customer value is driven by both purchase frequency and high-value transactions.
- One customer achieved an average order value of over 3,184 per order, highlighting the importance of premium purchases.

These customers represent strong candidates for loyalty programs, personalized marketing campaigns, and premium customer retention strategies.

In [105]:
pd.read_sql("""
SELECT
    strftime('%Y-%m', order_purchase_timestamp) AS purchase_month,
    ROUND(SUM(payment_value), 2) AS monthly_revenue,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master
GROUP BY purchase_month
ORDER BY purchase_month;
""", connection)

,purchase_month,monthly_revenue,total_orders
0,2016-09,388.47,4
1,2016-10,76559.05,324
2,2016-12,19.62,1
3,2017-01,190806.27,800
4,2017-02,351848.13,1780
5,2017-03,547769.84,2682
6,2017-04,512126.52,2404
7,2017-05,737425.31,3700
8,2017-06,613777.41,3245
9,2017-07,749242.84,4026


### Observation

Monthly revenue and order volume show a strong upward trend throughout 2017, indicating steady marketplace growth.

- Revenue generally increased alongside order volume, suggesting healthy business expansion.
- Minor fluctuations occurred in April and June, but the overall trajectory remained positive.
- December 2016 contains only one recorded order with minimal revenue, indicating a likely data collection limitation or incomplete monthly data rather than an actual business decline.

Overall, the marketplace experienced sustained growth in both customer demand and revenue over the observed period.

### Business Impact

The consistent increase in monthly revenue suggests that the marketplace successfully expanded its customer base and sales volume during 2017. Understanding the drivers behind peak months can help optimize inventory planning, marketing campaigns, and operational capacity.

# Advanced Analysis 1: Revenue by Product Category

## Business Question

Which product categories generate the highest revenue, and how can the business use this information to improve sales and profitability?

### Objective

Identify the highest-performing product categories based on:
- Total Revenue
- Number of Orders
- Average Order Value

This analysis helps identify the categories that contribute the most to overall business performance and supports decisions related to inventory planning, marketing investments, and pricing strategy.

pd.read_sql("""
SELECT
    strftime('%Y-%m', order_purchase_timestamp) AS purchase_month,
    ROUND(SUM(payment_value),2) AS monthly_revenue,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master
GROUP BY purchase_month
ORDER BY purchase_month;
""", connection)

In [106]:
pd.read_sql("""
SELECT
    strftime('%Y-%m', order_purchase_timestamp) AS purchase_month,
    ROUND(SUM(payment_value),2) AS monthly_revenue,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master
GROUP BY purchase_month
ORDER BY purchase_month;
""", connection)

,purchase_month,monthly_revenue,total_orders
0,2016-09,388.47,4
1,2016-10,76559.05,324
2,2016-12,19.62,1
3,2017-01,190806.27,800
4,2017-02,351848.13,1780
5,2017-03,547769.84,2682
6,2017-04,512126.52,2404
7,2017-05,737425.31,3700
8,2017-06,613777.41,3245
9,2017-07,749242.84,4026


# SQL Analysis 6: Customer Segmentation

## Business Question

Which customers contribute the most to the business, and how are customers distributed based on purchase frequency?

### Business Value

Customer segmentation helps businesses:

- Identify VIP customers
- Build loyalty programs
- Increase customer retention
- Personalize marketing campaigns
- Allocate marketing budgets more efficiently

In [107]:
pd.read_sql("""
SELECT
    CASE
        WHEN total_orders = 1 THEN 'One-time Customer'
        WHEN total_orders BETWEEN 2 AND 4 THEN 'Repeat Customer'
        ELSE 'Loyal Customer'
    END AS customer_segment,
    COUNT(*) AS customers,
    ROUND(AVG(total_orders),2) AS avg_orders
FROM (
    SELECT
        customer_unique_id,
        COUNT(DISTINCT order_id) AS total_orders
    FROM olist_master
    GROUP BY customer_unique_id
) customer_orders
GROUP BY customer_segment
ORDER BY customers DESC;
""", connection)

,customer_segment,customers,avg_orders
0,One-time Customer,93099,1.00
1,Repeat Customer,2978,2.09
2,Loyal Customer,19,6.47


## Business Insight

The extremely small proportion of loyal customers suggests that the business experiences high customer churn after the first purchase.

Potential strategies include:

- Loyalty and rewards programs
- Personalized email campaigns
- Product recommendations
- Subscription or membership programs
- Targeted promotions for repeat customers

Improving repeat purchase rates can substantially increase customer lifetime value without proportionally increasing customer acquisition costs.

# SQL Analysis 7: Product Category Performance

## Business Question

Which product categories generate the highest revenue, receive the most orders, and attract the largest customer base?

---

## Business Value

Understanding category performance helps businesses:

- Identify top-performing product categories
- Allocate inventory efficiently
- Optimize marketing spend
- Prioritize high-growth categories
- Improve assortment planning

In [108]:
pd.read_sql("""
SELECT
    product_category_name,
    ROUND(SUM(payment_value), 2) AS total_revenue,
    COUNT(DISTINCT order_id) AS total_orders,
    COUNT(DISTINCT customer_unique_id) AS unique_customers
FROM olist_master
GROUP BY product_category_name
ORDER BY total_revenue DESC
LIMIT 10;
""", connection)

,product_category_name,total_revenue,total_orders,unique_customers
0,cama_mesa_banho,1743998.80,9417,9145
1,beleza_saude,1662963.59,8836,8678
2,informatica_acessorios,1599481.06,6689,6557
3,moveis_decoracao,1443963.61,6449,6317
4,relogios_presentes,1430553.48,5624,5547
5,esporte_lazer,1400223.07,7720,7515
6,utilidades_domesticas,1097900.09,5884,5821
7,automotivo,855095.68,3897,3852
8,ferramentas_jardim,840721.59,3518,3489
9,cool_stuff,781933.97,3632,3615


## Observation

The Home, Bed & Bath (cama_mesa_banho) category generated the highest revenue, exceeding 1.74 million across more than 9,400 orders.

Beauty & Health and Computer Accessories closely followed, indicating strong customer demand across both lifestyle and technology products.

The top-performing categories consistently exhibit high revenue, high order volume, and a large unique customer base, suggesting these categories form the company's primary revenue drivers.

## Business Insight

The business should prioritize investment in high-performing categories such as Home, Bed & Bath and Beauty & Health through inventory optimization, supplier partnerships, and targeted marketing campaigns.

Technology-related categories also demonstrate strong customer demand, presenting opportunities for cross-selling and premium product offerings.

Lower-performing categories should be evaluated for profitability and inventory optimization to improve overall operational efficiency.

# SQL Analysis 8: Seller Performance Analysis

## Business Question

Which sellers generate the highest revenue and process the largest number of customer orders?

---

## Business Value

Seller performance analysis helps businesses:

- Identify top-performing sellers
- Build strategic partnerships
- Reward high-performing sellers
- Improve seller onboarding
- Optimize marketplace performance

In [109]:
pd.read_sql("""
SELECT
    seller_id,
    ROUND(SUM(price), 2) AS total_sales,
    COUNT(DISTINCT order_id) AS total_orders,
    ROUND(AVG(price), 2) AS avg_order_value
FROM olist_master
GROUP BY seller_id
ORDER BY total_sales DESC
LIMIT 10;
""", connection)

,seller_id,total_sales,total_orders,avg_order_value
0,53243585a1d6dc2643021fd1853d8905,244627.55,358,559.79
1,4869f7a5dfa277a7dca6462dcf3b52b2,237867.23,1132,200.56
2,4a3ca9315b744ce9f8e9374361493884,215825.77,1806,100.15
3,fa1c13f2614d7b5c4749cbc52fecda94,203984.22,585,334.95
4,7c67e1448b00f6e969d365cea6b010ab,199688.11,982,136.49
5,7e93a43ef30c4f03f38b393420bc753a,182878.17,336,519.54
6,da8622b14eb17ae2831f4ac5b9dab84a,171784.57,1314,103.36
7,7a67c85e85bb2ce8582c35f2203ad736,150749.79,1160,121.08
8,1025f0e2d44d7041d6cf58b6550e0bfa,143675.53,915,97.28
9,955fee9216a65b617aa5c0531780ce60,137405.00,1287,89.81


## Observation

The top-performing seller generated total sales of 244,627.55 from only 358 orders, achieving the highest average order value (559.79) among the top sellers.

Other leading sellers processed more than 1,000 orders but recorded lower average order values, indicating a high-volume sales strategy.

These results suggest that marketplace success is driven by both premium pricing and high transaction volume, depending on the seller's business model.

## Business Insight

The marketplace should adopt different engagement strategies for premium and high-volume sellers.

Premium sellers can be supported through exclusive promotions, premium placement, and long-term partnerships, while high-volume sellers may benefit from logistics optimization and inventory support.

Monitoring seller performance enables the business to strengthen relationships with strategic sellers while reducing dependence on any single supplier.

# SQL Analysis 9: Payment Method Analysis

## Business Question

Which payment methods are most frequently used by customers, and which contribute the highest revenue?

---

## Business Value

Understanding customer payment preferences helps businesses:

- Improve the checkout experience
- Optimize payment processing costs
- Increase transaction success rates
- Negotiate better terms with payment providers
- Design targeted promotional campaigns

In [110]:
pd.read_sql("""
SELECT
    payment_type,
    COUNT(*) AS total_transactions,
    ROUND(SUM(payment_value), 2) AS total_revenue,
    ROUND(AVG(payment_value), 2) AS avg_payment_value
FROM olist_master
GROUP BY payment_type
ORDER BY total_revenue DESC;
""", connection)

,payment_type,total_transactions,total_revenue,avg_payment_value
0,credit_card,87776,15775450.54,179.72
1,boleto,23190,4110920.74,177.27
2,voucher,6465,435917.84,67.43
3,debit_card,1706,257374.89,150.86
4,not_defined,3,0.00,0.00
5,NaN,3,NaN,NaN


# SQL Analysis 10: Customer Satisfaction Analysis

---

## Business Question

How does customer satisfaction, measured through review scores, impact order volume and revenue?

---

## Business Value

Customer review analysis helps businesses:

- Measure customer satisfaction
- Improve product and seller quality
- Detect operational issues affecting customer experience
- Increase customer retention
- Protect marketplace reputation

---

## SQL Query
```

In [111]:
pd.read_sql("""
SELECT
    CASE
        WHEN total_orders = 1 THEN 'One-time Customer'
        ELSE 'Repeat Customer'
    END AS customer_type,

    COUNT(*) AS customers,

    ROUND(AVG(total_orders), 2) AS avg_orders,

    ROUND(AVG(total_spend), 2) AS avg_customer_value

FROM (

    SELECT
        customer_unique_id,
        COUNT(DISTINCT order_id) AS total_orders,
        SUM(payment_value) AS total_spend

    FROM olist_master
    GROUP BY customer_unique_id

) customer_summary

GROUP BY customer_type
ORDER BY avg_customer_value DESC;
""", connection)

,customer_type,customers,avg_orders,avg_customer_value
0,Repeat Customer,2997,2.12,487.14
1,One-time Customer,93099,1.00,205.37


## Observation

The customer base is dominated by one-time purchasers, representing over 93,000 customers, while only 2,997 customers made repeat purchases.

Although repeat customers account for a much smaller share of the customer base, they place more than twice as many orders on average and generate an average customer value of 487.14 compared with 205.37 for one-time customers.

This demonstrates that repeat customers contribute significantly more value per customer and represent a critical driver of long-term revenue.

## Business Insight

The analysis highlights a substantial opportunity to improve customer retention.

Increasing the repeat purchase rate through loyalty programs, personalized recommendations, email marketing, and post-purchase engagement could significantly increase customer lifetime value without relying solely on acquiring new customers.

Customer retention should be monitored as a key business KPI alongside revenue growth and customer acquisition to support sustainable long-term profitability.

# Export Dataset for Power BI

The SQL analysis is complete. The final analytical dataset is exported as a CSV file for use in Power BI, where interactive dashboards and executive-level visualizations will be created.

In [112]:
powerbi_df = pd.read_sql("""
SELECT *
FROM olist_master
""", connection)

powerbi_df.to_csv(
    "data/cleaned/olist_powerbi.csv",
    index=False
)

print("Dataset exported successfully!")
print("Shape:", powerbi_df.shape)

OSError: Cannot save file into a non-existent directory: 'data/cleaned'

In [113]:
powerbi_df.to_csv(
    "../data/cleaned/olist_powerbi.csv",
    index=False
)

In [114]:
import os

print(os.path.exists("../data/cleaned/olist_powerbi.csv"))

True


In [115]:
powerbi_df.shape

(119143, 44)

In [116]:
powerbi_df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,review_creation_date,review_answer_timestamp,delivery_days,delivery_delay,total_order_value,shipping_ratio,order_month,order_year,order_quarter,purchase_weekday
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
1,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
2,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,2017-10-11 00:00:00,2017-10-12 03:43:48,8.0,-8.0,38.71,0.290764,October,2017,Q4,Monday
3,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,...,2018-08-08 00:00:00,2018-08-08 18:37:50,13.0,-6.0,141.46,0.191744,July,2018,Q3,Tuesday
4,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,2018-08-18 00:00:00,2018-08-22 19:07:58,9.0,-18.0,179.12,0.120200,August,2018,Q3,Wednesday


In [118]:
pd.read_sql("""
SELECT
    ROUND(SUM(payment_value), 2) AS total_revenue,
    COUNT(DISTINCT order_id) AS total_orders
FROM olist_master;
""", connection)

,total_revenue,total_orders
0,20579664.01,99441
